### DISCUSSION TITLE
How to reset history-index on CTRL-R with CTRL-C

### DISCUSSION THREAD

**Discussion by dezza** [akinomyoga/ble.sh#540] [Q&A] [answered]
Hello.

I've been looking thru `ble-bind` and the wiki to find a way to reset the cursor in the history entries. After using <kbd>Ctrl</kbd> + <kbd>R</kbd> the history-index will be decremented and won't reset if you press <kbd>Ctrl</kbd> + <kbd>C</kbd>. This means if you retrigger <kbd>Ctrl</kbd> + <kbd>R</kbd> then it will not start from the beginning.

I tried <kbd>Ctrl</kbd> + <kbd>G</kbd> which should be `isearch` or `nsearch/cancel`.

It seems like <kbd>Ctrl</kbd> + <kbd>J</kbd> or <kbd>Enter</kbd> does indeed reset it.

Is it possible to reset the history-index on <kbd>Ctrl</kbd> + <kbd>C</kbd> ?

**Comment by akinomyoga** [comment]
Use the following setting:

```bash
# blerc

ble-bind -m 'isearch' -f C-c isearch/cancel
```

----

By default <kbd>C-g</kbd> is the key to cancel the search. It reverts the history position to the original one (i.e., the one before starting the incremental search). To find out how to set it up so that <kbd>C-c</kbd> behaves the same as <kbd>C-g</kbd>, you can first look at the result of `ble-bind -m isearch -P` (Note that the search by <kbd>C-r</kbd> is the incremental search mode (`isearch`), and the keymap for `isearch` is named as `isearch`):

```bash
$ ble-bind -m isearch -P
ble-bind -m 'isearch' -f BS isearch/prev
ble-bind -m 'isearch' -f RET isearch/exit
ble-bind -m 'isearch' -f DEL isearch/prev
ble-bind -m 'isearch' -f __defchar__ isearch/self-insert
ble-bind -m 'isearch' -f __default__ isearch/exit-default
ble-bind -m 'isearch' -f __line_limit__ nop
ble-bind -m 'isearch' -f C-RET isearch/accept-line
ble-bind -m 'isearch' -f 'C-?' isearch/prev
ble-bind -m 'isearch' -f C-g isearch/cancel
ble-bind -m 'isearch' -f C-h isearch/prev
ble-bind -m 'isearch' -f C-j isearch/accept-line
ble-bind -m 'isearch' -f C-m isearch/exit
ble-bind -m 'isearch' -f C-r isearch/backward
ble-bind -m 'isearch' -f C-s isearch/forward
ble-bind -m 'isearch' -f 'C-x C-g' isearch/cancel
ble-bind -m 'isearch' -f C-M-g isearch/cancel
```

Then you'll find the line of `C-g`. Mimicking that line, you can write the keybinding setting for `C-c` in your `~/.blerc`. You can just replace `C-g` with `C-c`.

**Comment by akinomyoga** [comment]
Hmm, maybe I misunderstood what you meant.

Maybe it is actually unrelated to <kbd>C-r</kbd>, but it is related to the behavior of <kbd>C-c</kbd> outside the search mode? I think I can change the behavior of <kbd>C-c</kbd> (`discard-line`).

**Reply by dezza** [reply]
Ah yes, now I see `C-g` actually works, maybe I was stuck in a state before..

Yes indeed, the idea is to make `C-c` (discard-line) work as `C-g`

**Reply by akinomyoga** [reply, marked as answer]
I fixed the behavior of `discard-line` in commit 3b2b4b816ef0bacb4e161cf05f2b11cd09617559. @dezza Could you update `ble.sh` by running `ble-update` and check the behavior?

**Reply by dezza** [reply]
> I fixed the behavior of `discard-line` in commit 3b2b4b816ef0bacb4e161cf05f2b11cd09617559. @dezza Could you update `ble.sh` by running `ble-update` and check the behavior?
> 
> 

For sure I will, friend! Thanks a lot for your help! :)

**Reply by dezza** [reply]
Tested now. It works as intended :) Thanks again!

**Reply by akinomyoga** [reply]
@dezza OK! Thanks for checking the behavior!
